File: java/solitaire/klondike/src/main/java/org/silnith/game/solitaire/move/filter/MoveCapFilter.java
package org.silnith.game.solitaire.move.filter;

import java.util.List;

import org.silnith.game.GameState;
import org.silnith.game.solitaire.Board;
import org.silnith.game.solitaire.move.SolitaireMove;

/**
 * A move filter that caps the total game length.
 * 
 * <p>Realistically there is a finite limit to the number of moves it takes
 * to win a game of Klondike solitaire.  So searching a tree to a depth
 * greater than that number is unproductive.</p>
 */
public class MoveCapFilter implements SolitaireMoveFilter {

	private final int moveCap;
	private final String key;

	/**
	 * Creates a new move cap filter with the given cap.
	 * 
	 * @param moveCap the maximum number of moves to allow in a game
	 */
	public MoveCapFilter(final int moveCap) {
		super();
		this.moveCap = moveCap;
		this.key = "Move Cap of " + this.moveCap;
	}

	@Override
    public Object getStatisticsKey() {
        return key;
    }

    @Override
	public boolean shouldFilter(final List<GameState<SolitaireMove, Board>> gameStateHistory) {
		return gameStateHistory.size() > moveCap;
	}

}

File: csharp/Solitaire/Klondike/Move/Filter/MoveCapFilter.cs
using System.Collections.Generic;

namespace Silnith.Game.Klondike.Move.Filter
{
    /// <summary>
    /// A move filter that caps the total game length.
    /// </summary>
    /// <remarks>
    /// <para>
    /// Realistically there is a finite limit to the number of moves it takes
    /// to win a game of Klondike solitaire.  So searching a tree to a depth
    /// greater than that number is unproductive.
    /// </para>
    /// </remarks>
    public class MoveCapFilter : ISolitaireMoveFilter
    {
        /// <summary>
        /// The maximum number of moves to allow in a game.
        /// </summary>
        public int MoveCap
        {
            get;
        }

        /// <inheritdoc/>
        public object StatisticsKey
        {
            get;
        }

        /// <summary>
        /// Constructs a new move cap filter with the given cap.
        /// </summary>
        /// <param name="moveCap">The maximum number of moves to allow in a game.</param>
        public MoveCapFilter(int moveCap)
        {
            MoveCap = moveCap;
            StatisticsKey = "Move Cap of " + moveCap;
        }

        /// <inheritdoc/>
        public bool ShouldFilter(IReadOnlyList<GameState<ISolitaireMove, Board>> gameStateHistory)
        {
            return gameStateHistory.Count > MoveCap;
        }
    }
}
